launch_ros/launch_ros/actions/node.py

```diff
@@ -432,3 +481,86 @@ def expanded_node_namespace(self):
         return self.__expanded_remappings
+
+    def _get_extensions(self):
+        group_name = 'launch_ros.node_action'
+        entry_points = {}
+        for entry_point in importlib_metadata.entry_points().get(group_name, []):
+            entry_points[entry_point.name] = entry_point
+        extension_types = {}
+        for entry_point in entry_points:
+            try:
+                extension_type = entry_points[entry_point].load()
+            except Exception as e:  # noqa: F841
+                self.__logger.warning(
+                    f"Failed to load entry point '{entry_point.name}': {e}")
+                continue
+            extension_types[entry_points[entry_point].name] = extension_type
+
+        self.__extensions = {}
+        for extension_name, extension_class in extension_types.items():
+            extension_instance = self._instantiate_extension(
+                group_name, extension_name, extension_class, self.__logger)
+            if extension_instance is None:
+                continue
+            extension_instance.NAME = extension_name
+            self.__extensions[extension_name] = extension_instance
+
+    def _instantiate_extension(
+        self, group_name, extension_name, extension_class, logger, *, unique_instance=False
+    ):
+        if not unique_instance and extension_class in self.__extensions:
+            return self.__extensions[extension_name]
+
+        try:
+            extension_instance = extension_class()
+        except PluginException as e:  # noqa: F841
+            logger.warning(
+                f"Failed to instantiate '{group_name}' extension "
+                f"'{extension_name}': {e}")
+            return None
+        except Exception as e:  # noqa: F841
+            logger.error(
+                f"Failed to instantiate '{group_name}' extension "
+                f"'{extension_name}': {e}")
+            return None
+        if not unique_instance:
+            self.__extensions[extension_name] = extension_instance
+        return extension_instance
+
+
+class PluginException(Exception):
+    """Base class for all exceptions within the plugin system."""
+
+    pass
+
+
+def satisfies_version(version, caret_range):
+    assert caret_range.startswith('^'), 'Only supports caret ranges'
+    extension_point_version = Version(version)
+    extension_version = Version(caret_range[1:])
+    next_extension_version = get_upper_bound_caret_version(
+        extension_version)
+
+    if extension_point_version < extension_version:
+        raise PluginException(
+            'Extension point is too old (%s), the extension requires '
+            "'%s'" % (extension_point_version, extension_version))
+
+    if extension_point_version >= next_extension_version:
+        raise PluginException(
+            'Extension point is newer (%s), than what the extension '
+            "supports '%s'" % (extension_point_version, extension_version))
+
+
+def get_upper_bound_caret_version(version):
```

Comment:
@gbiggs consider moving these functions to another file.

Reply:
Done in 6754e9a366589dec1b1535eeef2e08ef765b1301.